### PR TITLE
S32 support for AEC

### PR DESCRIPTION
**Pull request by cujomalainey**

### CODE REVIEW

**Review by lgirdwood — COMMENTED** [review]
@singalsu @andrula-song pls review.

File: src/audio/google/google_rtc_audio_processing.c
Comment (singalsu) [inline src/audio/google/google_rtc_audio_processing.c:519]: Buffer_acquire() removal applies also for this usage, right?

**Review by singalsu — APPROVED** [review]

File: src/audio/google/google_rtc_audio_processing.c
Comment (lgirdwood) [reply src/audio/google/google_rtc_audio_processing.c:519]: yep, we can remove after if this merges first.

**Review by lgirdwood — COMMENTED** [review]
LGTM. @andrula-song pls also review.

**Review by lgirdwood — APPROVED** [review]
@cujomalainey good to merge. Some conflicts though.

**Comment by cujomalainey** [comment]
> @cujomalainey good to merge. Some conflicts though.

Negative, tests reported broken audio, working on debugging, just had to fight my tooling super hard this week

**Review by btian1 — COMMENTED** [review]
any performance difference with float api?

**Comment by cujomalainey** [comment]
> any performance difference with float api?

From what I understand, none, since the s16 is essentially a frontend for the f32 api

**Comment by cujomalainey** [comment]
This is blocked by #8258. The AEC xruns the pipe on ll

**Review by marcinszkudlinski — CHANGES_REQUESTED** [review]
Please do not merge, i'll include those changes in AEC DP processing PR

File: src/audio/google/google_rtc_audio_processing.c
Comment (btian1) [inline src/audio/google/google_rtc_audio_processing.c:685]: those Q_CONVERT_FLOAT(QTOF) macros normally used with fixed numbers, if you use here, you will have very negative performance impact, this is not proposed to use this macro, please find other way to convert to fix point data.

Comment (cujomalainey) [reply src/audio/google/google_rtc_audio_processing.c:685]: If there is a problem with the standard macros then they should be fixed or an alternative provided. The problem should not be addressed here if there is no other alternative currently.

**Comment by cujomalainey** [comment]
Test ping @thesofproject/google

**Comment by cujomalainey** [comment]
Sorry one more test, debugging team configs @thesofproject/google

File: src/audio/google/google_rtc_audio_processing.c
Comment (andyross) [inline src/audio/google/google_rtc_audio_processing.c:433]: Looks like there's a `!aec_reference_buffer` early exit case just below this, probably best to put this after that out of general "do the cheap checks first" hygiene.  It also allows the compiler (maybe) or you to roll this into the other num_capture_channels loop you're adding below.  I don't think I see any other branching here that would prevent that, though it's hard to tell in a review UI.

File: src/audio/google/google_rtc_audio_processing.c
Comment (andyross) [inline src/audio/google/google_rtc_audio_processing.c:444]: Style nitpick: this is a whitespace-only change, and it faked me out because I thought for a second because I was sure that this was changing the bzero() region without updating the size of the allocation.

IMHO the whole "write the size expression in the alloc and bzero steps" pattern is fragile.  I guess my aesthetics would say that if we're going to change the source code bytes that compute this number, we should fix it so the size is computed exactly once and used as an immutable value in both the alloc/clear.

File: src/audio/google/google_rtc_audio_processing.c
Comment (andyross) [inline src/audio/google/google_rtc_audio_processing.c:459]: Confused: this is looping over the same array as the num_aec_reference_channels loop above and initializing the same indexes to the same values.   Looks like semantically it's identical to looping once over the maximum of the two counts?  Or maybe it's supposed to be initializing a different array?

File: src/audio/google/google_rtc_audio_processing.c
Comment (andyross) [reply src/audio/google/google_rtc_audio_processing.c:685]: Yeah, unless I'm looking at the wrong code, it looks to me like this macro is a little pessimal, dividing (!) a float by a constructed 64 bit (!!) quantity (so at runtime this is going to emit a double precision division instead of the single precision multiply we want).  From format.h:

```
#define Q_CONVERT_QTOF(x, ny) ((float)(x) / ((int64_t)1 << (ny)))
```

Swapping that to something like this should generate better code for sure, though edge cases at the boundaries are always hard to reason about: `#define S32_TO_F(x)  ((x) * (1.0f / BIT(31)))`

File: src/audio/google/google_rtc_audio_processing.c
Comment (cujomalainey) [reply src/audio/google/google_rtc_audio_processing.c:459]: very likely a copy paste error

**Comment by cujomalainey** [comment]
> Notes and nitpicks. I guess one thing to note somewhere is that this is going to be highly dependent on platforms with a working FPU/HiFi engine and attendant toolchain and OS.

I don't think this makes a difference since this is essentially a no-op as the conversion to FP was already done inside the lib, this just makes its easily handed for multiple formats outside.

**Comment by cujomalainey** [comment]
No idea why github won't let me reply to https://github.com/thesofproject/sof/pull/8237#discussion_r1383779719 but for some reason there is no reply box.

this is a question for @ShriramShastry and @singalsu

**Comment by marcinszkudlinski** [comment]
@andyross @cujomalainey @singalsu  @nashif 
If its true that Zephyr does not save HiFi registers in preemption, it is a huge gap and must be fixed, DP processing is designed as a set of threads with deadlines, preempting each other (and being preempted by LL tasks)

**Comment by marcinszkudlinski** [comment]
As to slow float->int conversion, the answer is:
```
int16_t CONVERT(float data)
{
	const xtfloat ratio = 2 << 14;
	xtfloat x0 = data;
	xtfloat x1;
	int16_t x;

	x1 = XT_MUL_S(x0, ratio);
	x = XT_TRUNC_S(x1, 0);

	return x;
}
```

converting of 10ms of 4channel data from float to int16 took 1.1ms (!!!!) using Q_CONVERT_FLOAT, 120us using CONVERT as above
Q_CONVERT_QTOF, however, is quite fast, conversion from 10ms/4chanels fom int16 to float takes 140ms

**Comment by andyross** [comment]
Regarding HiFi registers: I'm guessing @dcpleung might have some WIP on saving the SIMD registers, if not I can take a crack at it.  It's not particularly difficult, though it does have some stack/performance tradeoffs to worry about (it's a ton of data to push on the stacks, but then stacks are cached where e.g. the thread struct can't be).  And IIRC (i.e. I might be wrong, would have to look up) Xtensa doesn't have a way to lazy-evaluate (e.g. trap on use) the context switch, we'd likely pay the cost for every interrupt/switch.

And as far as how to implement it: my recommendation would be to stay away from the explicit TIE macros as they defeat the optimizer (and obviously aren't portable anyway).  Checking quickly, the loop below (very roughly what the code in question is doing) already optimizes to a pleasingly unrolled loop over the scalar FPU with xcc 2021.6 on TGL.  With xt-clang and the ACE config, it actually emits some pretty scarily optimized (but quite fast looking) well-bundled HiFi SIMD loops.  The TIE macros are definitely in third place by my accounting (though still ahead of gcc: the Zephyr SDK for TGL won't emit FPU code at all and calls into libgcc -- pretty sure FPU support is supposed to work, might be a compiler build goof).

**Comment by andyross** [comment]
> Checking quickly, the loop below [...]

Oops, forgot loop:

```c
void f_to_s32(float *in, int *out, int n)
{
    for (int i = 0; i < n; i++)
        out[i] = in[i] * (1.0f / INT_MAX);
}
```

**Comment by andyross** [comment]
> And IIRC (i.e. I might be wrong, would have to look up) Xtensa doesn't have a way to lazy-evaluate (e.g. trap on use) the context switch

I looked it up.  It does.  It gets complicated with SMP in the presence of SCHED_PIN_ONLY=n because you need to shoot down the state on another core when threads migrate, but that's not a SOF configuration so we can defer optimizing that case.

File: src/audio/google/google_rtc_audio_processing.c
Comment (andrula-song) [inline src/audio/google/google_rtc_audio_processing.c:603]: same here.

**Comment by ShriramShastry** [comment]
> No idea why github won't let me reply to [#8237 (comment)](https://github.com/thesofproject/sof/pull/8237#discussion_r1383779719) but for some reason there is no reply box.
> 
> this is a question for @ShriramShastry and @singalsu


> Q_CONVERT_QTOF

Hi @cujomalainey ,

I think you can use `'Q_CONVERT_QTOF'` without any problems for the time being. In `cmocka` wrappers, `Q_CONVERT_QTOF` is frequently utilized, and there is no appreciable loss of accuracy in terms of Total Harmonic Distortion or Unit in last place.



If performance is an issue, I'm not sure what the expectations are—what the cycle and benefit are expected to be, how long they will last, and so on. 

Yes, it is true that the `#define Q_CONVERT_QTOF(x, ny) ((float)(x) / ((int64_t)1 << (ny)))` division operator may be an argument. In that case, the implementation might be made quickly for any compiler by introducing the inverse `#define Q_CONVERT_QTOF(x, ny) ((float)(x) * inv((int64_t)1 << (ny)))` and employing a leading-to-zeros technique with a LUT to prevent loops.

Thank you!

**Comment by marcinszkudlinski** [comment]
I'm finishing with DP venison of AEC and the loop looks like:

```
	for (int i = 0; i < cd->num_frames; i++) {
		for (channel = 0; channel < cd->num_capture_channels; channel++)
			cd->process_buffer_ptrs[channel][i] =
					Q_CONVERT_QTOF(src[channel], 15);

		src += cd->num_capture_channels;
		if (src >= src_buf_end)
			src = src_buf_start;
	}
```
execution time for the loop above - for 10ms 16bit 4 channels - 140us


**please do not merge this commit - once DP version is stable - hopefully bvvery soon - i'll put a new PR for this**

**Comment by marcinszkudlinski** [comment]
I';m closing this PR, 
all changes for 32bit are here: https://github.com/thesofproject/sof/pull/8469

**Pull request closed without merging**
